src/edx/chain_table.rs

```diff
@@ -0,0 +1,187 @@
+//! Implements the Chain Table algorithm.
+//!
+//! This algorithm is in charge of storing the lists of values indexed by
+//! Findex. Formally, it implements an Encrypted Dictionary (EDX) scheme.
+//!
+//! The encryption scheme used is AES256-GCM.
+
+use std::{
+    collections::{HashMap, HashSet},
+    ops::{Deref, DerefMut},
+};
+
+use cosmian_crypto_core::{kdf256, reexport::rand_core::CryptoRngCore, SymmetricKey};
+
+use crate::{
+    edx::{
+        structs::{EdxKey, Seed},
+        DxEnc, EdxStore,
+    },
+    error::Error,
+    parameters::{SEED_LENGTH, TOKEN_LENGTH},
+    EncryptedValue, Label,
+};
+
+/// Chain Table representation.
+pub struct ChainTable<const VALUE_LENGTH: usize, Edx: EdxStore<VALUE_LENGTH>>(pub Edx);
+
+impl<const VALUE_LENGTH: usize, Edx: EdxStore<VALUE_LENGTH>> Deref
+    for ChainTable<VALUE_LENGTH, Edx>
+{
+    type Target = Edx;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+const CHAIN_TABLE_KEY_DERIVATION_INFO: &[u8] = b"Chain Table key derivation info.";
+
+impl<
+    const VALUE_LENGTH: usize,
+    EdxScheme: EdxStore<VALUE_LENGTH, EncryptedValue = EncryptedValue<VALUE_LENGTH>>,
+> DxEnc<VALUE_LENGTH> for ChainTable<VALUE_LENGTH, EdxScheme>
+{
+    type EncryptedValue = EncryptedValue<VALUE_LENGTH>;
+    type Error = Error<EdxScheme::Error>;
+    type Key = EdxKey;
+    type Seed = Seed<SEED_LENGTH>;
+    type Store = EdxScheme;
+    type Token = EdxScheme::Token;
+
+    fn setup(edx: Self::Store) -> Self {
+        Self(edx)
+    }
+
+    fn gen_seed(&self, rng: &mut impl CryptoRngCore) -> Self::Seed {
+        Seed::new(rng)
+    }
+
+    fn derive_keys(&self, seed: &Self::Seed) -> Self::Key {
```

Comment:
When deriving keys, why do you pass constants in parameters  ? `CHAIN_TABLE_KEY_DERIVATION_INFO` and `b"KMAC key"` ? Why do you need to update the hash with public constants ?

Reply:
These constants are used as domain separation information. Different constants are used for the Entry Table.

Reply:
Oh yes, because you derive two times from the same initial key. Didn't see, makes sense :+1: